maven-core/src/main/java/org/apache/maven/execution/LockingEventSpy.java

```diff
@@ -0,0 +1,96 @@
+package org.apache.maven.execution;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.maven.eventspy.AbstractEventSpy;
+import org.apache.maven.eventspy.EventSpy;
+import org.apache.maven.project.MavenProject;
+import org.eclipse.aether.SessionData;
+import org.eclipse.sisu.Typed;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * EventSpy implementation that provides a per-project locking mechanism
+ * to make sure a given project can not be build twice concurrently.
+ * This case can happen when running parallel builds with forked lifecycles.
+ */
+@Singleton
+@Named
+@Typed( EventSpy.class )
+public class LockingEventSpy extends AbstractEventSpy
+{
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    private static final Object LOCKS_KEY = new Object();
+
+    @SuppressWarnings( { "unchecked", "rawtypes" } )
+    private Lock getLock( ExecutionEvent event )
+    {
+        SessionData data = event.getSession().getRepositorySession().getData();
+        Map<MavenProject, Lock> locks = ( Map ) data.get( LOCKS_KEY );
+        if ( locks == null )
+        {
+            data.set( LOCKS_KEY, null, new ConcurrentHashMap<>() );
```

Comment:
What mean `null` here?
Three parameters?
This line can be called concurrently. It should be treated somehow, like the semantics `trySet(): boolean`

Reply:
I don't think there's a concurrency issue here.
The code does the following:
```
        Map<MavenProject, Lock> locks = ( Map ) data.get( LOCKS_KEY );
        if ( locks == null )
        {
            data.set( LOCKS_KEY, null, new ConcurrentHashMap<>() );
            locks = ( Map ) data.get( LOCKS_KEY );
        }
```
The `data.set(k, null, v)` performs a `map.putIfAbsent(k, v)`, so this can not overwrite the value.  In case there is a concurrent access to this very method, the following line does a `get`, so I think there's no way the code can use a wrong map.

Reply:
ok, if the session is implemented using `putIfAbsent` then it should be fine.